# 3. Indexing, slicing

Each element of an array can be located by its position in each dimension. Numpy offers multiple ways to access single elements or groups of elements in very efficient ways. We will illustrate these concepts both with small simple matrices as well as a regular image, in order to illustrate them.

In [1]:
import numpy as np

**3.1 Accessing single values**

We create a small 2D array to use as an example:




In [2]:
normal_array = np.random.normal(10, 2, (3,4))
normal_array

array([[12.95404643,  8.72733769,  7.61469081,  9.67748218],
       [13.72186234, 11.17773204,  7.90484459,  8.75604287],
       [10.03777463, 13.46756129, 10.35170442, 11.14855255]])

It is very easy to access an array's values. One can just pass an *index* for each dimensions. For example to recover the value on the last row and second column of the ```normal_array``` array we just write (remember counting starts at 0):

In [3]:
single_value = normal_array[2,1]
single_value

9.268641173173533

What is returned in that case is a single number that we can re-use:

In [4]:
single_value += 10
single_value

19.268641173173535

And that change doesn't affect the original value in the array:

In [5]:
normal_array

array([[ 7.90771964,  9.83585098,  6.11253703,  9.85337274],
       [ 9.05131344,  8.90320551,  8.80594885, 11.98047351],
       [ 9.22904886,  9.26864117, 11.95082707, 10.60275012]])

However we can also directly change the value in an array:

In [6]:
normal_array[2,1] = 23

In [7]:
normal_array

array([[ 7.90771964,  9.83585098,  6.11253703,  9.85337274],
       [ 9.05131344,  8.90320551,  8.80594885, 11.98047351],
       [ 9.22904886, 23.        , 11.95082707, 10.60275012]])

## 3.2 Accessing part of an array with indices: slicing

### 3.2.1 Selecting a range of elements

One can also select multiple elements in each dimension (e.g. multiple rows and columns in 2D) by using the ```start:end:step``` syntax. By default, if omitted, ```start=0```, ```end=last element``` and ```step=1```. For example to select the first **and** second rows of the first column, we can write:

In [22]:
# row 0 ra 1 ko column 0 ma vako value dinxa

In [23]:
normal_array[0:2,0]

array([7.90771964, 9.05131344])

In [24]:
normal_array

array([[ 7.90771964,  9.83585098,  6.11253703,  9.85337274],
       [ 9.05131344,  8.90320551,  8.80594885, 11.98047351],
       [ 9.22904886, 23.        , 11.95082707, 10.60275012]])

In [25]:
normal_array[0:4,1]

array([ 9.83585098,  8.90320551, 23.        ])

Note that the ```end``` element is **not** included. One can use the same notation for all dimensions:

In [26]:
normal_array[0:2,2:4]

array([[ 6.11253703,  9.85337274],
       [ 8.80594885, 11.98047351]])

In [27]:
normal_array[1:,2:4]

array([[ 8.80594885, 11.98047351],
       [11.95082707, 10.60275012]])

### 3.2.2 Selecting all elements
If we only specify ```:```, it means we want to recover all elements in that dimension:

In [28]:
normal_array[:,2:4]

array([[ 6.11253703,  9.85337274],
       [ 8.80594885, 11.98047351],
       [11.95082707, 10.60275012]])

Also in general, if you only specify the value for a single axis, this will take the first element of the first dimension:

In [29]:
normal_array

array([[ 7.90771964,  9.83585098,  6.11253703,  9.85337274],
       [ 9.05131344,  8.90320551,  8.80594885, 11.98047351],
       [ 9.22904886, 23.        , 11.95082707, 10.60275012]])

In [30]:
normal_array[1]

array([ 9.05131344,  8.90320551,  8.80594885, 11.98047351])

Finally note that if you want to recover only one element along a dimension (single row, column etc), you can do that in two ways:

In [31]:
normal_array[0,:]

array([7.90771964, 9.83585098, 6.11253703, 9.85337274])

This returns a one-dimensional array containing a single row from the original array:

In [32]:
normal_array[0,:].shape

(4,)

Instead, if you specify actual boundaries that still return only a single row:

In [33]:
normal_array[0:1,:]

array([[7.90771964, 9.83585098, 6.11253703, 9.85337274]])

kashyap ghimire

In [34]:
normal_array[0:1,:].shape #new dimension is added.

(1, 4)

### The difference in the results of normal_array[0:1, :] and normal_array[0:1] is due to how slicing and indexing work in NumPy arrays, particularly when specifying rows and columns.

1) normal_array[0:1, :]:
0:1 means "select rows starting from index 0 up to but not including index 1."
: means "select all columns."
This operation explicitly specifies both the row range and the column range.
Since both the row and column ranges are specified, the result is a 2D array:
ie. array([[7.90771964, 9.83585098, 6.11253703, 9.85337274]])
2) normal_array[0:1]:
0:1 means "select rows starting from index 0 up to but not including index 1."
Since only the row range is specified and the column range is omitted, NumPy automatically assumes that all columns are included.
In this case, NumPy simplifies the result because it's taking the first slice in the first dimension (rows) and including all columns. However, since only the row range is explicitly given and not the column range, NumPy reduces the result to a 1D array:

## 3.2 Sub-arrays are not copies!

As often with Python when you create a new variable using a sub-array, that variable **is not independent** from the original variable:

In [36]:
sub_array = normal_array[:,2:4]

In [37]:
sub_array

array([[ 6.11253703,  9.85337274],
       [ 8.80594885, 11.98047351],
       [11.95082707, 10.60275012]])

In [38]:
normal_array

array([[ 7.90771964,  9.83585098,  6.11253703,  9.85337274],
       [ 9.05131344,  8.90320551,  8.80594885, 11.98047351],
       [ 9.22904886, 23.        , 11.95082707, 10.60275012]])

If for example we modify ```normal_array```, this is going to be reflected in ```sub_array``` too:

In [39]:
normal_array[0,2] = 100

In [40]:
normal_array

array([[  7.90771964,   9.83585098, 100.        ,   9.85337274],
       [  9.05131344,   8.90320551,   8.80594885,  11.98047351],
       [  9.22904886,  23.        ,  11.95082707,  10.60275012]])

In [41]:
sub_array

array([[100.        ,   9.85337274],
       [  8.80594885,  11.98047351],
       [ 11.95082707,  10.60275012]])

The converse is also true:

In [42]:
sub_array[0,1] = 50

In [43]:
sub_array

array([[100.        ,  50.        ],
       [  8.80594885,  11.98047351],
       [ 11.95082707,  10.60275012]])

In [44]:
normal_array

array([[  7.90771964,   9.83585098, 100.        ,  50.        ],
       [  9.05131344,   8.90320551,   8.80594885,  11.98047351],
       [  9.22904886,  23.        ,  11.95082707,  10.60275012]])

If you want your sub-array to be an *independent* copy of the original, you have to use the ```.copy()``` method:

In [45]:
sub_array_copy = normal_array[1:3,:].copy()

In [46]:
sub_array_copy

array([[ 9.05131344,  8.90320551,  8.80594885, 11.98047351],
       [ 9.22904886, 23.        , 11.95082707, 10.60275012]])

In [47]:
sub_array_copy[0,0] = 500

In [48]:
sub_array_copy

array([[500.        ,   8.90320551,   8.80594885,  11.98047351],
       [  9.22904886,  23.        ,  11.95082707,  10.60275012]])

In [49]:
normal_array

array([[  7.90771964,   9.83585098, 100.        ,  50.        ],
       [  9.05131344,   8.90320551,   8.80594885,  11.98047351],
       [  9.22904886,  23.        ,  11.95082707,  10.60275012]])

## 3.4. Accessing parts of an array with coordinates

In the above case, we are limited to select rectangular sub-regions of the array. But sometimes we want to recover a series of specific elements for example the elements (row=0, column=3) and (row=2, column=2). To achieve that we can simply index the array with a list containing row indices and another with columns indices:

In [52]:
row_indices = [0,2]
col_indices = [3,2]

normal_array[row_indices, col_indices]

array([50.        , 11.95082707])

In [53]:
normal_array

array([[  7.90771964,   9.83585098, 100.        ,  50.        ],
       [  9.05131344,   8.90320551,   8.80594885,  11.98047351],
       [  9.22904886,  23.        ,  11.95082707,  10.60275012]])

In [54]:
selected_elements = normal_array[row_indices, col_indices]

In [55]:
selected_elements

array([50.        , 11.95082707])

## 3.5 Logical indexing

The last way of extracting elements from an array is to use a boolean array of same shape. For example let's create a boolean array by comparing our original matrix to a threshold:

In [56]:
bool_array = normal_array > 40
bool_array

array([[False, False,  True,  True],
       [False, False, False, False],
       [False, False, False, False]])

We see that we only have two elements which are above the threshold. Now we can use this logical array to *index* the original array. Imagine that the logical array is a mask with holes only in ```True``` positions and that we superpose it to the original array. Then we just take all the values visible in the holes:

In [57]:
normal_array[bool_array]

array([100.,  50.])

## 3.6 Reshaping arrays

Often it is necessary to reshape arrays, i.e. keep elements unchanged but change their position. There are multiple functions that allow one to do this. The main one is of course ```reshape```.

### 3.6.1 ```reshape```

Given an array of $MxN$ elements, one can reshape it with a shape $OxP$ as long as $M*N = O*P$.

In [58]:
reshaped = np.reshape(normal_array,(2,6))
reshaped

array([[  7.90771964,   9.83585098, 100.        ,  50.        ,
          9.05131344,   8.90320551],
       [  8.80594885,  11.98047351,   9.22904886,  23.        ,
         11.95082707,  10.60275012]])

In [59]:
reshaped.shape

(2, 6)

In [60]:
reshaped1 = np.reshape(normal_array,(12,))
reshaped1

array([  7.90771964,   9.83585098, 100.        ,  50.        ,
         9.05131344,   8.90320551,   8.80594885,  11.98047351,
         9.22904886,  23.        ,  11.95082707,  10.60275012])

In [61]:
np.reshape(normal_array,(4,3))

array([[  7.90771964,   9.83585098, 100.        ],
       [ 50.        ,   9.05131344,   8.90320551],
       [  8.80594885,  11.98047351,   9.22904886],
       [ 23.        ,  11.95082707,  10.60275012]])

### 3.6.2 Flattening

It's also possible to simply flatten an array i.e. remove all dimensions to create a 1D array. This can be useful for example to create a histogram of a high-dimensional array.

In [62]:
flattened = np.ravel(normal_array)
flattened

array([  7.90771964,   9.83585098, 100.        ,  50.        ,
         9.05131344,   8.90320551,   8.80594885,  11.98047351,
         9.22904886,  23.        ,  11.95082707,  10.60275012])

In [63]:
flattened.shape

(12,)

### 3.6.3 Dimension collapse

Another common way that leads to reshaping is projection. Let's consider again our ```normal_array```:

In [64]:
normal_array

array([[  7.90771964,   9.83585098, 100.        ,  50.        ],
       [  9.05131344,   8.90320551,   8.80594885,  11.98047351],
       [  9.22904886,  23.        ,  11.95082707,  10.60275012]])

We can project all values along the first or second axis, to recover for each row/column the largest value:

In [65]:
proj0 = np.max(normal_array, axis = 0)
proj0

array([  9.22904886,  23.        , 100.        ,  50.        ])

In [66]:
proj0.shape

(4,)

### 3.6.4 Swaping dimensions

We can also simply exchange the position of dimensions. This can be achieved in different ways. For example we can ```np.roll``` dimensions, i.e. circularly shift dimensions. This conserves the relative oder of all axes:

In [67]:
array3D = np.ones((4, 10, 20))
array3D.shape

(4, 10, 20)

In [68]:
array3D

array([[[1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
         1., 1., 1., 1.]],

       [

In [69]:
array_rolled = np.rollaxis(array3D, axis=1, start=0)
array_rolled.shape

(10, 4, 20)

Alternatively you can swap two axes. This doesn't preserver their relative positions:

In [72]:
array_swapped = np.swapaxes(array3D, 0,2)
array_swapped.shape

(20, 10, 4)

In [87]:
# visualize rolling


In [88]:
o=np.array([[[1,2,3,4],
           [5,6,7,8],
           [9,10,11,12]],
          [[13,14,15,16],
          [17,18,19,20],
          [21,22,23,24]]])

In [89]:
o.shape

(2, 3, 4)

In [86]:
np.rollaxis(o,1).shape

(3, 2, 4)

In [90]:
np.rollaxis(o,2)

array([[[ 1,  5,  9],
        [13, 17, 21]],

       [[ 2,  6, 10],
        [14, 18, 22]],

       [[ 3,  7, 11],
        [15, 19, 23]],

       [[ 4,  8, 12],
        [16, 20, 24]]])

In [91]:
np.rollaxis(o,2).shape

(4, 2, 3)

### 3.6.5 Change positions

Finally, we can also change the position of elements without changing the shape of the array. For example if we have an array with two columns, we can swap them:

In [92]:
array2D = np.random.normal(0,1,(4,2))
array2D

array([[ 0.65826354, -0.28309991],
       [ 0.74809104,  0.10049608],
       [-0.81125457, -1.00471895],
       [-1.51096325,  1.68668415]])

In [93]:
np.fliplr(array2D)

array([[-0.28309991,  0.65826354],
       [ 0.10049608,  0.74809104],
       [-1.00471895, -0.81125457],
       [ 1.68668415, -1.51096325]])

Similarly, if we have two rows:

In [94]:
array2D = np.random.normal(0,1,(2,4))
array2D

array([[ 1.01347047, -0.49262378,  0.35023213,  0.69955314],
       [-0.58754846, -0.36818165,  0.46528785, -0.3143481 ]])

In [95]:
np.flipud(array2D)

array([[-0.58754846, -0.36818165,  0.46528785, -0.3143481 ],
       [ 1.01347047, -0.49262378,  0.35023213,  0.69955314]])